### DISCUSSION TITLE
How to transition between child and parent states in HSM?

### DISCUSSION THREAD

**Discussion by drvelociraptor** [limbonaut/limboai#249] [Q&A] [answered]
Basically I have a movement state and an attack state as a child. I want to change from movement to attack and viceversa. Can this be done? It seems add_transition just works with direct child states, and change_active_state gives me this error: "Unable to perform transition to a state that is not a child of this HSM."... Don't know what to do because I can't find anything on the docs, all examples seem to be for basic FSM.

**Comment by limbonaut** [comment]
Can you provide an example of your state hierarchy?
And correct - transitions can be declared for the direct children only. Nested HSM can have their own transitions. (Nested) HSM has `initial_state` which dictates which substate becomes active when it is entered. If not set, it defaults to the first child.

**Reply by drvelociraptor** [reply]
Hi, thanks for the response. Basically my state hierarchy is as described. I add to the root a state called "terra", wich refers to the ground movement, and add a child state to it called "atackTF".  At the beginning of the game i'm on "terra". I just want to change the state to "atackTF", and back to "terra" after an animation is played. I need this because when grounded you'll have access to various attacks that also follow the grounded rules. So I assume the "terra" update funciton will also run when we are on a "terra" child state? I'm very novice in gamedev in general so I may be missing something obious here. I attach an image of the remote scene view.
![imagen](https://github.com/user-attachments/assets/0e99b405-89cb-4202-9011-cab8fa700457)

**Reply by limbonaut** [reply]
Yes, the `_update` function is called for all active states. As long as one of the substates is active, the parent state is also considered active, and thus its function is executed as well.

In your example, the nested HSM is active only if one of its substates is active. If the "terra" state is currently active, exactly one of its child states must be active too. By design, when a nested HSM becomes active, its `initial_state` is also activated. Transitioning is the process of changing an active substate within an HSM (or nested HSM). 

Based on the explanation above, you cannot transition directly from the "atacTF" state to the "terra" state, as they are not at the same level. Following this logic, transitioning to the "terra" state would activate the "atacTF" state (presumably the **initial** state, as it is the first state in the node tree). If you want to maintain the logic of movement within the parent state, you could add a "NoAttack" or "NoAction" state that delegates control to the parent state. Then, you can declare a transition from the "NoAction" state to the "AttackTF" state as follows:
```gdscript
terra.add_transition(no_action, attack_tf, "some_event")
```
Calling `dispatch("some_event")` anywhere in the state tree would trigger that transition, as long as the "NoAction" state is currently active.

**Reply by drvelociraptor** [reply]
I see, then if I understand correctly, you can only have an active state with no children? Like the deepest possible one in the tree, which I believe is called a 'leaf' in LimboAI? I assume there’s a reason for this, but can I ask why it works this way? Wouldn't it be better to allow being in specific states rather than just in a leaf state? It seems limiting and redundant to add empty states just to mimic being in the parent state. But yeah, my specific case is probably not the general one. Anyway, thank you for your work and the detailed explanations, I’m really liking this tool.

**Reply by drvelociraptor** [reply]
Just tryied to add_transition to terra and this happens: "Invalid call. Nonexistent function 'add_transition' in base 'LimboState'." In your example "terra" was a LimboState?

**Reply by limbonaut** [reply]
> Just tryied to add_transition to terra and this happens: "Invalid call. Nonexistent function 'add_transition' in base 'LimboState'." In your example "terra" was a LimboState?

I think, "terra" should be LimboHSM, and in your case it is LimboState

**Reply by limbonaut** [reply]
State machine is not just in "leaf" state. It is in each state that is currently active, so in case of active `atacTF` it would be in the following states: LimboHSM, terra, atacTF. However, terra should be LimboHSM type, not LimboState. LimboState can't have child states.

**Reply by limbonaut** [reply, marked as answer]
> I assume there’s a reason for this, but can I ask why it works this way?

`LimboState` is not a composite state, so it cannot contain other states as children (you could use class inheritance instead if you want to share some functionality). Adding more states under vanilla `LimboState` not going to work. And `LimboHSM` is a composite that can have exactly one child substate active. We call it active substate. Think of it like a switch - only one of the children states is allowed to be active. And this is what transitions do - switch the active substate on a level. `LimboHSM` is a state itself (inherits from `LimboState`), so you can nest them like so:
```
- LimboHSM1
--- LimboHSM2
----- LimboState1
----- LimboState2
--- LimboState
```
So in other words,
- `LimboState` is a leaf
- `LimboHSM` is a container (aka composite), can be nested

`LimboHSM` can define its own logic as well, so if `LimboState1` is active, `LimboHSM2` is also active, so is `LimboHSM1`. But by definition, HSM must be in one of its substates.

You can read more here: https://statecharts.dev/glossary/compound-state.html
We are not exactly following state charts, but in this case, most details are pretty much how we do it.

**Reply by drvelociraptor** [reply]
Thanks for the explanation! It all makes sense now. States cannot have childs and HSM act like state folders. Here the correct structure that works for me now: 
![imagen](https://github.com/user-attachments/assets/b500d981-12b1-44f8-b58e-baba4f3ae000)
